### PR TITLE
[DependencyInjection] added a way to reference dynamic environment variables in service definitions

### PR DESCRIPTION
**Pull request by fabpot**
| Q | A |
| --- | --- |
| Bug fix? | yesish |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #10138, #7555 |
| License | MIT |
| Doc PR | not yet |

When hosting a Symfony app on some PaaS like Heroku, or even if you are just using Docker, you need a way to reference environment variables in the container. That's already possible in Symfony but the values are dumped into the cache. This PR adds a way to reference an env variable that is resolved dynamically even on a compiled container.

ping @dzuelke

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/EnvVariable.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    /**
+     * __toString.
+     *
+     * @return string The environment variable name
+     */
+    public function __toString()
+    {
+        return (string) $this->id;
```
Comment (dosten) [inline src/Symfony/Component/DependencyInjection/EnvVariable.php]: IMO is better to cast to a string in the constructor.

**Comment by dzuelke** [comment]
It looks like this still dumps literal values into the container instead of `getenv()` calls:

I'm using your branch:

```
~ $ grep getenv vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php              
        return sprintf("getenv('%s')", $name);
```

My `app/config/config_prod.yml` has this (and I moved the base DB config to `config_dev.yml`):

```
doctrine:
    dbal:
        # instead of configuring the database access options in this file, we pull
        # them from the app/config/parameters.yml file. The reason is that config.yml
        # stores options that change the application behavior and parameters.yml
        # stores options that change from one server to another
        url:   "$DATABASE_URL$"
        charset:  UTF8
```

But the container has a literal value:

```
~ $ grep postgres app/cache/prod/appProdProjectContainer.php
        return $this->services['doctrine.dbal.default_connection'] = $this->get('doctrine.dbal.connection_factory')->createConnection(array('url' => 'postgres://someuser:somepass@ec2-107-21-221-59.compute-1.amazonaws.com:5432/somedatabase', 'charset' => 'UTF8', 'host' => 'localhost', 'port' => NULL, 'user' => 'root', 'password' => NULL, 'driver' => 'pdo_mysql', 'driverOptions' => array()), new \Doctrine\DBAL\Configuration(), new \Symfony\Bridge\Doctrine\ContainerAwareEventManager($this), array());
```

If I remember correctly, that's the exact problem I had when I tried to implement this myself. I was looking for a solution where there's a `getenv()` call in the generated container, and that would have been possible for Symfony's own stuff, but not for third-party bundles with their own dumper, as they or their calling code would have to be aware of this new syntax. There seems to be no central way of injecting this.

Or am I doing something wrong? I used the symfony-demo project with a few `dev` requirements for all the `symfony/polyfill-…` repositories.

File: src/Symfony/Component/DependencyInjection/EnvVariable.php (hunk 1 above)
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/EnvVariable.php]: or dont cast at all. The parameter is already defined as string.

**Comment by fabpot** [comment]
@dzuelke That's intended (or at least, that's the way I see it): a parameter IS evaluated, an env variable is NOT. So, you should use an env variable directly in your definitions instead of assigning the env variable to a parameter first, in which case, it's evaluated. That might be confusing though, so the second commit does what you expected. The only "problem" I can see is that this new behavior introduces some inconsistencies: a parameter storing an env variable is not evaluated only when used as a constructor argument or a method call argument; the same parameter used for a class name, or an alias for instance will be evaluated (as the code expects static values for those elements, and I don't want to support dynamic values here). WDYT?

**Comment by sstok** [comment]
If you only use it for service definitions you can simple use a custom Expression function.
And that would also be less of a BC break :+1:

**Comment by fabpot** [comment]
Which BC break?

**Comment by iltar** [comment]
I think where a parameter is actually named $FOO$, not entirely sure though, never had the use-case myself.

**Comment by fabpot** [comment]
Using an expression is indeed a very good idea and should work well. I'll have later.

**Comment by dzuelke** [comment]
> So, you should use an env variable directly in your definitions instead of assigning the env variable to a parameter first

How would I do that, @fabpot?

> If you only use it for service definitions you can simple use a custom Expression function.

That's what @ruudk suggested in https://github.com/symfony/symfony/issues/7555#issuecomment-99968935 and what @stof implemented in https://github.com/Incenteev/DynamicParametersBundle but the problem AFAICT is that every bundle out there would have to be modified to use that.

**Comment by WouterJ** [comment]
Can't we change the syntax to `$SOMETHING`? Env variable names cannot include whitespaces, so it should not cause any problems (and this syntax is equal to how people are used to reference env variables in shell/bash scripts).

File: src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBag.php [hunk 2]
```diff
@@ -167,12 +168,12 @@ public function resolve()
      * @throws ParameterCircularReferenceException if a circular reference if detected
      * @throws RuntimeException                    when a given parameter has a type problem.
      */
-    public function resolveValue($value, array $resolving = array())
+    public function resolveValue($value, array $resolving = array(), $protectEnvVariables = false)
```
Comment (ogizanagi) [inline src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBag.php]: The new `protectEnvVariables` param is missing from the docblock.

File: src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBag.php [hunk 3]
```diff
@@ -197,7 +198,7 @@ public function resolveValue($value, array $resolving = array())
      * @throws ParameterCircularReferenceException if a circular reference if detected
      * @throws RuntimeException                    when a given parameter has a type problem.
      */
-    public function resolveString($value, array $resolving = array())
+    public function resolveString($value, array $resolving = array(), $protectEnvVariables = false)
```
Comment (ogizanagi) [inline src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBag.php]: Same here

**Comment by fabpot** [comment]
@dzuelke Can you test the latest version of this patch? It should work fine now.

**Comment by dzuelke** [comment]
Sure @fabpot, will do. I thought you were going to rewrite it to use expressions first, my bad :)

File: src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBag.php [hunk 4]
```diff
@@ -236,7 +245,7 @@ public function resolveString($value, array $resolving = array())
             $resolved = (string) $resolved;
             $resolving[$key] = true;
 
-            return $self->isResolved() ? $resolved : $self->resolveString($resolved, $resolving);
+            return $self->isResolved() ? $resolved : $self->resolveString($resolved, $resolving, $protectEnvVariables);
```
Comment (dzuelke) [inline src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBag.php]: The variable is missing from the `use()` statement!

**Comment by dzuelke** [comment]
The `preg_replace_callback` references `$protectEnvVariables` but does not `use()` it, @fabpot:

```
> post-update-cmd: Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::clearCache


                                                             
  [Symfony\Component\Debug\Exception\ContextErrorException]  
  Notice: Undefined variable: protectEnvVariables            
                                                             


Script Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::clearCache handling the post-update-cmd event terminated with an exception


                                                                             
  [RuntimeException]                                                         
  An error occurred when executing the "'cache:clear --no-warmup'" command.  
                                                                             


Exception trace:
 () at /Users/dzuelke/Code/heroku/php-test/symfony_env/symfony-demo/vendor/sensio/distribution-bundle/Sensio/Bundle/DistributionBundle/Composer/ScriptHandler.php:439
 Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::executeCommand() at /Users/dzuelke/Code/heroku/php-test/symfony_env/symfony-demo/vendor/sensio/distribution-bundle/Sensio/Bundle/DistributionBundle/Composer/ScriptHandler.php:138
 Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::clearCache() at phar:///usr/local/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php:211
 Composer\EventDispatcher\EventDispatcher->executeEventPhpScript() at phar:///usr/local/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php:167
 Composer\EventDispatcher\EventDispatcher->doDispatch() at phar:///usr/local/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php:92
 Composer\EventDispatcher\EventDispatcher->dispatchScript() at phar:///usr/local/bin/composer/src/Composer/Installer.php:349
 Composer\Installer->run() at phar:///usr/local/bin/composer/src/Composer/Command/UpdateCommand.php:143
 Composer\Command\UpdateCommand->execute() at phar:///usr/local/bin/composer/vendor/symfony/console/Command/Command.php:256
 Symfony\Component\Console\Command\Command->run() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:838
 Symfony\Component\Console\Application->doRunCommand() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:189
 Symfony\Component\Console\Application->doRun() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:147
 Composer\Console\Application->doRun() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:120
 Symfony\Component\Console\Application->run() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:82
 Composer\Console\Application->run() at phar:///usr/local/bin/composer/bin/composer:43
 require() at /usr/local/bin/composer:25
```

**Comment by dzuelke** [comment]
But even then, still no luck, @fabpot:

```
~ $ grep -Hn postgres app/cache/prod/appProdProjectContainer.php
app/cache/prod/appProdProjectContainer.php:340:        return $this->services['doctrine.dbal.default_connection'] = $this->get('doctrine.dbal.connection_factory')->createConnection(array('url' => 'postgres://someuser:somepass@ec2-107-21-221-59.compute-1.amazonaws.com:5432/somedatabase', 'charset' => 'UTF8', 'host' => 'localhost', 'port' => NULL, 'user' => 'root', 'password' => NULL, 'driver' => 'pdo_mysql', 'driverOptions' => array()), new \Doctrine\DBAL\Configuration(), new \Symfony\Bridge\Doctrine\ContainerAwareEventManager($this), array());
```

I'm using a branch based on yours, with my fix (dzuelke/symfony@1cf6e985b3a3b3071ea022293141e2d6704f1cb9) applied:

```
~ $ grep protectEnvVariables vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBag.php | wc -l
8
```

**Comment by dzuelke** [comment]
Also paging @stof for thoughts. Or am I doing something really wrong here?

**Comment by dzuelke** [comment]
I also played around with various permutations around parameters, and `parameters.yml`, where I have:

```
# This file is auto-generated during the composer install
parameters:
    database_driver: pdo_sqlite
    database_host: 127.0.0.1
    database_port: null
    database_name: null
    database_user: root
    database_password: null
    database_path: '%kernel.root_dir%/data/blog.sqlite'
    mailer_transport: smtp
    mailer_host: 127.0.0.1
    mailer_user: null
    mailer_password: null
    locale: en
    secret: secret_value_for_symfony_demo_application
    database_url: $DATABASE_URL$
```

The value there also gets written out to the cache literally and not as a `getenv()` call.

**Comment by dzuelke** [comment]
Ping @fabpot / @stof

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 5]
```diff
@@ -1377,6 +1380,8 @@ private function dumpValue($value, $interpolate = true)
                 // we do this to deal with non string values (Boolean, integer, ...)
                 // the preg_replace_callback converts them to strings
                 return $this->dumpParameter(strtolower($match[1]));
+            } elseif (preg_match('/^\$([^\$]+)\$$/', $value, $match)) {
```
Comment (sstok) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: This is will also match `$foo'boom$` making it possible to unwillingly inject PHP code.

And the $$ will not match but is not normalized to a single $ in the else-block.

File: src/Symfony/Component/DependencyInjection/Dumper/YamlDumper.php [hunk 6]
```diff
@@ -262,6 +263,8 @@ private function dumpValue($value)
             return $code;
         } elseif ($value instanceof Reference) {
             return $this->getServiceCall((string) $value, $value);
+        } elseif ($value instanceof EnvVariable) {
+            return sprintf('$%s$', (string) $value);
```
Comment (sstok) [inline src/Symfony/Component/DependencyInjection/Dumper/YamlDumper.php]: The `escape` method should be updated to also escape a value starting with '$', no?

**Comment by lolautruche** [comment]
Hello

Problem with this format is that it clashes with eZ dynamic settings which uses the same. 
See https://doc.ez.no/display/EZP/Dynamic+settings+injection

**Comment by fabpot** [comment]
Conclusion: This is not possible, and by far. The semantic configuration used by bundle extensions is resolved before being passed to the extension. If we do not resolve it anymore, each extension would have to resolve values on a case-by-case basis AND would need to keep the non-resolved value. That might be a problem when a parameter has influences over the configuration.

So, a generic system is just not possible without rethinking the whole system.

The only thing I can think of is to allow the user to override some service arguments after the container has been compiled, but that would mean relying on conventions sometimes (when the service using the configuration has been dynamically created for instance).

Another possibility is to only add support for some bundles like the DSN for the Doctrine bundle. That might be easier and would probably fix most of the issues.

Closing for now.

**Comment by lolautruche** [comment]
More or less in the same field, it's been a while that we, at eZ, think about contributing our siteaccess system to enable one to have context aware settings (e.g.Multisite inside one app). 
Would this make sense in Symfony? Or should we better just externalize it and propose it as a lib/bundleBundle?

**Pull request closed without merging**